core/src/main/scala/cats/TraverseFilter.scala

```diff
@@ -203,4 +205,21 @@ object TraverseFilter {
   @deprecated("Use cats.syntax object imports", "2.2.0")
   object nonInheritedOps extends ToTraverseFilterOps
 
+  private[cats] def traverseFilterDirectly[Coll[x] <: IterableOnce[x], G[_], A, B](
+    builder: mutable.Builder[B, Coll[B]]
+  )(fa: IterableOnce[A])(f: A => G[Option[B]])(implicit G: StackSafeMonad[G]): G[Coll[B]] = {
+    val size = fa.knownSize
+    if (size >= 0) {
+      builder.sizeHint(size)
+    }
+    G.map(fa.iterator.foldLeft(G.pure(builder)) { case (bldrG, a) =>
+      G.flatMap(bldrG) { bldr =>
+        G.map(f(a)) {
+          case Some(b) => bldr += b
```

Comment:
I can't see why we know this is safe or lawful.

We have a mutable data structure that could potentially be in a multithreaded situation with G.

Also, consider cases like IO where you have a long computation that finally fails, then you recover some part of it to succeed. It feels like this mutable builder could remember things from failed branches.

I think using an immutable builder (like for instance just building up a List, Chain or Vector) would be much easier to verify it is lawful.

Reply:
> We have a mutable data structure that could potentially be in a multithreaded situation with G.

It can't be: every append to the builder happens in `flatMap` so by law it must be sequential, not concurrent/parallel.

> Also, consider cases like IO where you have a long computation that finally fails, then you recover some part of it to succeed. It feels like this mutable builder could remember things from failed branches.

Hmm. I'm not entirely sure how this applies to `traverse`. It doesn't have a notion of "recovery". For sure, each individual step that the `traverse` runs may have a notion of recovery, but that's just it: it will either succeed or it will fail. But there's no way to "recover" an intermediate structure from the `traverse` itself.

> I think using an immutable builder (like for instance just building up a List, Chain or Vector) would be much easier to verify it is lawful.

Won't disagree here. We could do a benchmark to see how much performance we are leaving on the table with that strategy.